### PR TITLE
`Lectures`: Allow instructors to import existing lectures into courses

### PR DESCRIPTION
**Pull request by ge65cer**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added `@PreAuthorize` and checked the course groups for all new REST Calls (security).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

Instructors usually hold Courses in a regular interval with little changes to the lecture content (https://github.com/ls1intum/Artemis/issues/4880).
Therefore, we introduce a process to import existing lectures from other Courses to the Course.

### Description

Using the import functionality, instructors can browse all Lectures they have access to.
In the modal, they can search and sort existing Lectures.
When clicking the import button, Artemis clones the Lecture including its Lecture Units and Attachments to the Course.
Exercise Units are excepted here, as there exists a separate process to import the different types of exercises on the Exercise Management page.
> This is one of my first pull requests on this project, so constructive feedback is very welcome.

### Steps for Testing

Prerequisites:
- 1 Instructor or Editor

1. Log in to Artemis
2. Navigate to Course Administration
3. Open a Course or create one where you are 
4. Create a Lecture, optionally with different Lecture Units and Attachments to the Lecture
5. Open or create another Course
6. Click the "Import Lecture" button on the Lecture Management page of that Course
7. Import a Lecture and verify Artemis also cloned the respective Lecture Units and Attachments

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Branch | Line |
|------------|-------:|-----:|
| LectureImportService.java | 92% | 99% |
| LectureService.java | 85% | 95% |
| LectureResource.java | 79% | 97% |
| lecture-import.component.ts  | 76% | 92% |

### Screenshots

screencast

### CODE REVIEW

**Review by bassner — CHANGES_REQUESTED** [review]
I've tested this on TS1 and it works perfectly fine. 👍 

The client code you added looks good. However, you didn't provide any tests for the new client components and services you added. I suggest to add some tests for them!

**Review by Hialus — CHANGES_REQUESTED** [review]
Reviewed the code.
Client code looks good tp me, but I left some comments for the server code.
You also did not add an integration test for your new endpoints. Please add them for both the search and the import lecture endpoint.

File: src/main/webapp/app/lecture/lecture.service.ts [hunk 1]
```diff
@@ -84,6 +84,22 @@ export class LectureService {
             );
     }
 
+    import(courseId: number, lectureId: number): Observable<EntityResponseType> {
```
Comment (Hialus) [inline src/main/webapp/app/lecture/lecture.service.ts:87]: Please add a JSDoc here.

**Review by Hialus — APPROVED** [review]
Thanks for applying my feedback.
Server, Client, and Test code look good to me now.
Well done 👍🏻

**Review by Hialus — DISMISSED** [review]
New test changes look good.
Reapprove Code.

File: src/test/javascript/spec/component/lecture/lecture-import.component.spec.ts [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+            declarations: [
+                LectureImportComponent,
+                MockComponent(ButtonComponent),
+                MockComponent(NgbHighlight),
+                MockComponent(NgbPagination),
+                MockDirective(SortByDirective),
+                MockDirective(SortDirective),
+            ],
+            providers: [],
+            schemas: [],
```
Comment (bassner) [inline src/test/javascript/spec/component/lecture/lecture-import.component.spec.ts:38]:
```suggestion
```

File: src/test/javascript/spec/component/lecture/lecture-import.component.spec.ts [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    });
+
+    afterEach(() => {
+        jest.restoreAllMocks();
+    });
+
+    it('should initialize', () => {
+        fixture.detectChanges();
+        expect(LectureImportComponent).not.toBeNull();
+    });
```
Comment (bassner) [inline src/test/javascript/spec/component/lecture/lecture-import.component.spec.ts:58]: Tests like this are, altough some of them exist in our codebase, useless. Please remove it :)

**Review by bassner — CHANGES_REQUESTED** [review]
Two minor nitpicks :)

File: src/test/javascript/spec/component/lecture/lecture-import.component.spec.ts [hunk 4]
```diff
@@ -0,0 +1,124 @@
+import { ComponentFixture, fakeAsync, TestBed, tick } from '@angular/core/testing';
+import { FormsModule } from '@angular/forms';
+import { NgbHighlight, NgbPagination } from '@ng-bootstrap/ng-bootstrap';
+import { ButtonComponent } from 'app/shared/components/button.component';
+import { SortService } from 'app/shared/service/sort.service';
+import { PageableSearch, SearchResult, SortingOrder } from 'app/shared/table/pageable-table';
+import { MockComponent, MockDirective, MockPipe } from 'ng-mocks';
```
Comment (Strohgelaender) [inline src/test/javascript/spec/component/lecture/lecture-import.component.spec.ts:7]: unused import (code warning)
```suggestion
import { MockComponent, MockDirective } from 'ng-mocks';
```

Comment (ge65cer) [reply src/test/javascript/spec/component/lecture/lecture-import.component.spec.ts:7]: Great to see my IDE doing its job… 🤖

**Review by bassner — DISMISSED** [review]
Thanks for adding the changes.
Looks good to me now! 👍 
Good work :)

**Review by maximiliansoelch — DISMISSED** [review]
Tested on TS1, works as expected 👍

**Review by Hialus — DISMISSED** [review]
Re-approve Code 👍🏻

File: src/main/java/de/tum/in/www1/artemis/service/LectureImportService.java [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+        attachment.setReleaseDate(importedAttachment.getReleaseDate());
+        attachment.setVersion(importedAttachment.getVersion());
+        attachment.setAttachmentType(importedAttachment.getAttachmentType());
+
+        Path oldPath = Paths.get(fileService.actualPathForPublicPath(importedAttachment.getLink()));
+        Path tempPath = Paths.get(FilePathService.getTempFilePath(), oldPath.getFileName().toString());
+
+        try {
+            log.debug("Copying attachment file from {} to {}", oldPath, tempPath);
+            Files.copy(new FileInputStream(oldPath.toFile()), tempPath, StandardCopyOption.REPLACE_EXISTING);
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/LectureImportService.java:158]:
```suggestion
            Files.copy(oldPath, tempPath, StandardCopyOption.REPLACE_EXISTING);
```

**Review by Strohgelaender — APPROVED** [review]
Code aproval

**Comment by ge65cer** [comment]
Another iteration, here we go… 🎡

**Review by bassner — APPROVED** [review]
Only minor things changed since my last review.
Reapprove 👍

**Review by Hialus — APPROVED** [review]
Code still looks good.
Re-approve 👍🏻

**Review by maximiliansoelch — APPROVED** [review]
Reapprove 👍

**Pull request merged**